# Nivell 1

## Exercici 1: Selecció de Personal

Tu treballes com a analista de dades en el departament de RRHH d'una empresa. S'ha obert una nova oferta laboral interna i l'empresa ha decidit que només les primeres 10 persones que s'inscriguin podran participar en el procés de selecció.
Per això, t'han encarregat desenvolupar una eina senzilla en Python que actuï com a formulari d'inscripció i sistema de classificació inicial. Aquest formulari ha de permetre introduir, per a cada candidat/a:

- Nom i Cognoms
- Telèfon de contacte
- Anys d'experiència dins de l'empresa
- Coneixement de Català (0–3 punts)
    - 0 = sense coneixement
    - 1 = fins a nivell B1
    - 2 = nivell B2
    - 3 = C1 o superior
- Domini d'Excel (0–3 punts)
    - 0 = nivell bàsic
    - 1 = ús de funcions
    - 2 = taules dinàmiques
    - 3 = Visual Basic i Macros
- Títol acadèmic (0–3 punts)
    - 0 = ESO
    - 1 = FP
    - 2 = Universitat
    - 3 = Màster o superior

Assegurant-se que tota la informació introduïda és correcta i entra dins els valors permesos. A partir d'aquestes dades, el programa haurà de calcular una puntuació total aplicant el pes de l'experiència (anys * 1.5) i sumant-hi els punts de català, excel i titulació, i a continuació classificar automàticament cada persona segons els llindars establerts per l'empresa:

» Prioritat Alta → puntuació ≥ 12

» Candidat adequat → 8 ≤ puntuació < 12

» No prioritari → puntuació < 8

Cada vegada que algú complet a el formulari, s'ha de mostrar la seva puntuació individual. Quan ja s'hagin inscrit exactament deu persones, l'eina ha de deixar d'acceptar nous registres i generar un informe final en pantalla, ordenat primer segons la prioritat i, en cas d'empat, per la puntuació obtinguda. Aquest informe ha d'incloure el nom, el telèfon i la puntuació de cada candidat per tal que el departament de RRHH els pugui contactar ràpidament.

<br>

 **Important**

Per resoldre aquest exercici, hauràs de:

» Desenvolupar un sistema modular amb les funcions següents:
- obtenir_dades
- validar_dades
- calcular_puntuacio
- classificar_candidat
- mostrar_individual
- mostrar_informe
- main_seleccio # orquestra totes les funcions
» Escriure primer l'algoritme pas a pas, indicant clarament quines operacions farà cada funció.
» Escriure el codi de cada funció.

<br>

Per fer proves:

```
inputs_test = [
    "Laura Serra", "608112233", 6, 3, 3, 3,
    "Clara Gómez", "655221199", 3, 3, 3, 2,
    "Sergio López", "644112211", 3, 2, 2, 2,
    "Candidat Error", "611000000", 5, 7, 3, 2, # Català = 7 → NO VÀLID
    "Marta Ruiz", "699887766", 2, 3, 2, 1,
    "Marc Vidal", "612443322", 6, 3, 3, 3,
    "Júlia Romero", "677998877", 5, 3, 3, 2,
    "Pol Navarro", "699554433", 4, 3, 2, 2,
    "David Ferrer", "677665544", 2, 1, 1, 1,
    "Anna Paredes", "633221144", 1, 2, 1, 0,  
    "Anibal Paredes", "644272727", 1, 2, 1, 0,  
    "Núria Casals", "600112200", 8, 2, 3, 3 # 11è candidat     
]
```



L'eina seguirà el següent fluxe:

1. Obtenció de dades mitjançant la funció obtenir_dades().
2. Validació de que els camps no estiguin buits mitjançant la funció validar_buits().
3. Conversió dels tipus de dades numèriques mitjançant la funció convertir_tipus().
4. Validació del tipus de dades introduïdes mitjançant la funció validar_tipus().
5. Validació dels rangs de les dades introduïdes mitjançant la funció validar_rang().
6. Càlcul de la puntuació de cada candidat a partir de les dades introduïdes mitjançant la funció calcular_puntuacio().
7. Classificació de la puntuació calculada en calcular_puntuació() mitjançant la funció classificar_candidat().
8. Mostrar l'informe del candidat, mitjançant mostrar_resultat().
9. Ordenar la prioritat dels resultats de la puntuació, mitjançant ordenar_prioritat().
10. Mostrar un informe final amb tots els candidats, ordenats per ordre de puntuació.
11. Executar l'eina amb main_seleccio().


#### obtenir_dades():
1. Demana a l'usuari les dades personals nom, telefon, anys.
2. Demana a l'usuari el nivell respectiu de català, excel i títol acadèmic.
3. Retorna un diccionari amb les dades.

In [1]:
def obtenir_dades():
    """
    Demana a l'usuari les dades per inscriure un candidat.
    Retorna un diccionari amb les dades.
    """

    nom = input("Nom i cognoms: ")
    telefon = input("Telèfon de contacte: ")
    anys = input("Anys d'experiència dins de l'empresa: ")

    print("\nConeixement de català (0–3 punts):")
    print(" 0 = Sense coneixement")
    print(" 1 = Fins a nivell B1")
    print(" 2 = -nivell B2")
    print(" 3 = C1 o superior")
    catala = input("Introdueix un valor (0-3): ")

    print("\nDomini d'Excel (0–3 punts):")
    print(" 0 = Nivell bàsic")
    print(" 1 = Ús de funcions")
    print(" 2 = Taules dinàmiques")
    print(" 3 = Visual Basic i Macros")
    excel = input("Introdueix un valor (0-3): ")

    print("\nTítol acadèmic (0–3 punts):")
    print(" 0 = ESO")
    print(" 1 = FP")
    print(" 2 = Universitat")
    print(" 3 = Màster o superior")
    titol = input("Introdueix un valor (0-3): ")

    return {
        "nom": nom,
        "telefon": telefon,
        "anys": anys,
        "catala": catala,
        "excel": excel,
        "titol": titol
    }

#### validar_buits():
1. Recull el diccionari obtingut a obtenir_dades().
2. Retorna False si hi ha algún camp buit.
3. Retorna True si tots els camps tenen algun valor.

In [2]:
def validar_buits(dades):
    """
    Valida que cap camp estigui buit.
    Retorna True si tots els camps tenen algun valor.
    """
    for camp, valor in dades.items():
        if valor == "":
            print(f"Error: el camp '{camp}' no pot estar buit.")
            return False
    return True


#### convertir_tipus():
1. Recull el diccionari obtingut a obtenir_dades().
2. Genera una llista amb els camps numèrics.
3. Converteix els valors dels camps numèrics a valors enters.
4. Retorna el diccionari amb els tipus de valors modificats.

In [3]:
def convertir_tipus(dades):
    """
    Converteix els camps numèrics del diccionari a enters.
    Retorna el diccionari modificat si tot és correcte.
    """
    camps_numerics = ["anys", "catala", "excel", "titol"]

    for camp in camps_numerics:
        try:
            dades[camp] = int(dades[camp])
        except ValueError:
            print(f"Error: el camp '{camp}' ha de ser un número vàlid.")
            return False

    return dades

#### validar_tipus():
1. Recull el diccionari generat per main_seleccio().
2. Valida que nom i telefon son de tipus str.
3. Retorna True si tot és correcte.

In [4]:
def validar_tipus(dades):
    """
    Valida el tipus de dades introduïdes.
    Retorna True si és correcte.
    """
    esquema = {
        "nom": str,
        "telefon": str
    }

    for camp, tipus_esperat in esquema.items():
        valor = dades.get(camp)
        if not isinstance(valor, tipus_esperat):
            print(f"Error: el camp '{camp}' ha de ser {tipus_esperat.__name__}")
            return False

    return True

#### validar_rang():
1. Recull les dades retornades per convertir_tipus().
2. Genera una llista amb els valors dels camps catala, excel i titol.
3. Valida que el valor de cada camp estigui dins del rang indicat. Si està fora del rang envia un missatge d'error i s'atura la validació.
4. Retorna True si tot és correcte.

In [5]:
def validar_rang(dades):
    """
    Valida el rang de les dades catala, excel, titol.
    Retorna True si és correcte.
    """
    camps = ["catala", "excel", "titol"]

    for camp in camps:
        valor = dades.get(camp)
        if not (0 <= valor <= 3):
            print(f"Error: el camp '{camp}' ha de ser un nombre entre 0 i 3")
            return False

    return True

#### calcular_puntuacio():
1. Recull els camps anys, catala, excel i titol.
2. Multiplica els anys d'experiència per 1.5.
3. Suma els valors de catala, excel, titol i el resultat de la multiplicació anterior.
4. Retorna la puntuació final (la suma anterior).

In [6]:
def calcular_puntuacio(anys, catala, excel, titol):
    """
    Calcula la puntuació total del candidat.
    Retorna la puntuació total final de cada candidat.
    """
    puntuacio_experiencia = anys * 1.5
    puntuacio_total = puntuacio_experiencia + catala + excel + titol
    return puntuacio_total

#### classificar_candidat():
1. Recull la puntuació retornada per calcular_puntuació().
2. Classifica la puntuació a partir dels varems (igual o superior a 12, igual o superior a 8 o menor de 8).
3. Retorna segons la classificació una valoració.

In [7]:
def classificar_candidat(puntuacio):
    """
    Classifica el candidat segons la puntuació total.
    """
    if puntuacio >= 12:
        return "Prioritat Alta"
    elif puntuacio >= 8:
        return "Candidat adequat"
    else:
        return "No prioritari"

#### mostrar_resultat():
1. Recull el diccionari generat per main_seleccio().
2. Mostra per pantalla les dades nom, telefon, puntuació i classificació del candidat.

In [8]:
def mostrar_resultat(candidat):
    """
    Mostra per pantalla la informació individual d'un candidat.
    """
    print("\n**** Resultat candidat/da ****")
    print(f"Nom: {candidat['nom']}")
    print(f"Telèfon: {candidat['telefon']}")
    print(f"Puntuació: {candidat['puntuacio']:.2f}")
    print(f"Classificació: {candidat['classificacio']}")
    print("******************************")

#### ordenar_prioritat():
1. Recull la classificació obtinguda a classificar_candidat().
2. Ordena la prioritat dels resultats de menor a major.
3. Retorna la prioritat i la puntuació del candidat.

In [9]:
def ordenar_prioritat(candidat):
    """
    Defineix l'ordre de prioritat per a l'informe final.
    """
    prioritat = candidat["classificacio"]

    if prioritat == "Prioritat Alta":
        ordre = 0
    elif prioritat == "Candidat adequat":
        ordre = 1
    else:
        ordre = 2

    return(ordre, -candidat["puntuacio"])

#### mostrar_informe():
1. Recull la llista generada a main_seleccio().
2. Ordena la llista per ordre de prioritat.
3. Mostra en pantalla les dades de la llista.

In [10]:
def mostrar_informe(llista_candidats):
    """
    Mostra l'informe final amb tots els candidats ordenats per la puntuació de major a menor.
    """
    print("\n****** INFORME FINAL ******")
    llista_ordenada = sorted(llista_candidats, key=ordenar_prioritat)

    for c in llista_ordenada:
        print(f"Nom: {c['nom']}")
        print(f"Telèfon: {c['telefon']}")
        print(f"Puntuació: {c['puntuacio']:.2f}")
        print(f"Classificació: {c['classificacio']}")
        print("------------------------")

    print("****** FI INFORME ******\n")

#### main_seleccio():
1. Genera una llista de candidats.
2. Si les dades introduides son incorrectes envia un missatge per pantalla demanant la re-introducció de les dades.
3. Recull la puntuació retornada per calcular_puntuació.
4. Crida validar_buits().
5. Crida convertir_tipus().
6. Crida validar_tipus().
7. Crida validar_rang().
8. Recull la classificació retornada per classificar_candidat.
9. Genera un diccionari per cada candidat amb el nom, telefon, puntuació i classificació.
10. Afegeix el candidat a la llista de candidats.
11. Executa mostra el resultat per cada candidat.
12. Un cop s'han introduit 10 candidats, informa per pantalla que no es poden introduir més dades de candidats.
13. Mostra l'informe final.


In [11]:
def main_seleccio():
    """
    Ordena tot el procés d'inscripció i classificació.
    """
    candidats = []

    while len(candidats) < 10:
        print(f"\n--- Formulari candidat/a {len(candidats) + 1} ---")

        while True:
            dades = obtenir_dades()

            if not validar_buits(dades):
                print("Si us plau, torna a introduir les dades.\n")
                continue

            dades_convertides = convertir_tipus(dades)
            if not dades_convertides:
                print("Si us plau, torna a introduir les dades.\n")
                continue

            if not validar_tipus(dades_convertides):
                print("Si us plau, torna a introduir les dades.\n")
                continue

            if not validar_rang(dades_convertides):
                print("Si us plau, torna a introduir les dades.\n")
                continue

            break

        puntuacio = calcular_puntuacio(
            dades_convertides["anys"],
            dades_convertides["catala"],
            dades_convertides["excel"],
            dades_convertides["titol"]
        )

        classificacio = classificar_candidat(puntuacio)

        candidat = {
            "nom": dades_convertides["nom"],
            "telefon": dades_convertides["telefon"],
            "puntuacio": puntuacio,
            "classificacio": classificacio
        }

        candidats.append(candidat)

        mostrar_resultat(candidat)

    print("\nJa s'han inscrit 10 persones. No s'accepten més candidats.\n")

    mostrar_informe(candidats)


#### main_seleccio():
1. Crida la funció main_seleccio().

In [12]:
main_seleccio()


--- Formulari candidat/a 1 ---
Nom i cognoms: Laura Serra
Telèfon de contacte: 608112233
Anys d'experiència dins de l'empresa: 6

Coneixement de català (0–3 punts):
 0 = Sense coneixement
 1 = Fins a nivell B1
 2 = -nivell B2
 3 = C1 o superior
Introdueix un valor (0-3): 3

Domini d'Excel (0–3 punts):
 0 = Nivell bàsic
 1 = Ús de funcions
 2 = Taules dinàmiques
 3 = Visual Basic i Macros
Introdueix un valor (0-3): 3

Títol acadèmic (0–3 punts):
 0 = ESO
 1 = FP
 2 = Universitat
 3 = Màster o superior
Introdueix un valor (0-3): 3

**** Resultat candidat/da ****
Nom: Laura Serra
Telèfon: 608112233
Puntuació: 18.00
Classificació: Prioritat Alta
******************************

--- Formulari candidat/a 2 ---
Nom i cognoms: Clara Gómez
Telèfon de contacte: 655221199
Anys d'experiència dins de l'empresa: 3

Coneixement de català (0–3 punts):
 0 = Sense coneixement
 1 = Fins a nivell B1
 2 = -nivell B2
 3 = C1 o superior
Introdueix un valor (0-3): 3

Domini d'Excel (0–3 punts):
 0 = Nivell bà

### Comentari final
Aquesta eina permet registrar candidats i obtenir-ne una classificació automàtica segons els criteris definits. Facilita la feina del departament de Recursos Humans, ja que estandarditza l’avaluació, redueix errors humans i assegura que tots els candidats es valoren amb els mateixos paràmetres. El resultat final depèn de les dades introduïdes, però el procés garanteix coherència i objectivitat en la selecció.

---
---



## Exercici 2: Segmentació per risc de cistella de compra

Treballes com a analista de dades al departament de Màrqueting Digital. L'empresa vol millorar l'efectivitat de les campanyes de recuperació de cistellas de compra abandonats i necessita identificar quins clients tenen més risc d'abandonament, quin segment de comportament presenten i quin missatge publicitari és el més adequat per a cadascun.
Disposes d'un conjunt de dades amb informació bàsica de 10 clients:

- id_client
- nom
- edat
- grup_edat
- codi_comportament
                
```
clients = [
    {"id_client": 1, "nom": "Laura Serra", "edat": 23, "grup_edat": "18-25", "codi_comportament": 4},
    {"id_client": 2, "nom": "Marc Vidal", "edat": 35, "grup_edat": "26-40", "codi_comportament": 3},
    {"id_client": 3, "nom": "Ana López", "edat": 29, "grup_edat": "26-40", "codi_comportament": 2},
    {"id_client": 4, "nom": "Joan Riera", "edat": 19, "grup_edat": "18-25", "codi_comportament": 2},
    {"id_client": 5, "nom": "Mónica Pérez", "edat": 47, "grup_edat": "41-60", "codi_comportament": 4},
    {"id_client": 6, "nom": "Luis García", "edat": 52, "grup_edat": "41-60", "codi_comportament": 1},
    {"id_client": 7, "nom": "Pilar Sánchez", "edat": 61, "grup_edat": "60+", "codi_comportament": 3},
    {"id_client": 8, "nom": "Eva Martín", "edat": 38, "grup_edat": "26-40", "codi_comportament": 0},
    {"id_client": 9, "nom": "Diego Romero", "edat": 24, "grup_edat": "18-25", "codi_comportament": 1},
    {"id_client": 10, "nom": "Núria Costa", "edat": 33, "grup_edat": "26-40", "codi_comportament": 4}]
```
             
El departament de CRM ha definit manualment:

» Codis de comportament
- 0 Navega però no afegeix productes
- 1 Afegeix productes però no arriba al checkout
- 2 Arriba al checkout però no paga
- 3 Cistella de compra abandonat fa menys de 24 h
- 4 Cistella de compra abandonat fa més de 24 h
» Factors segons codi de comportament
- Codi 0 0.10
- Codi 1 0.25
- Codi 2 0.50
- Codi 3 0.75
- Codi 4 0.90
» Factors segons grup d'edat
- 18–25 → 1.20
- 26–40 → 1.00
- 41–60 → 0.85
- 60+ → 0.70
» Fórmula del risc final

```risc_final = factor codi * factor edat```

Un cop processats els clients, has de ser capaç de respondre:

» 1. Quins clients s'han de contactar primer en una campanya de retargeting?

» 2. Quines diferències hi ha entre clients amb el mateix comportament però d'edats diferents? (Per exemple: un client jove que abandona la cistella de compra fa 24 h té el mateix risc que un de més edat?)

» 3. Quins clients presenten tan poc interès que no val la pena invertir-hi pressupost?

» 4. Quin client seria el més rendible de recuperar?
Interpretant el risc final i el segment.

<br>

**⚙️ Requisits tècnics del teu programa**
Hauràs de construir un sistema modular utilitzant diverses funcions i, sobretot, sense utilitzar if/elif/else per fer la segmentació.

<br>

 **Important**

Per resoldre aquest exercici, hauràs de:

» Desenvolupar un sistema modular amb les funcions següents:

» S'ha d'implementar amb un diccionari de diccionaris, sense condicions.
- processar_clients   # s'ha d'alimentar de les funcions següents:
    - obtenir_segment_i_factor
    - obtenir_factor_edat
    - calcular_risc
- mostrar_dataframe o mostrar_diccionari
- mostrar_respostes
- main_marketing    # orquestra totes les funcions

» Escriure primer l'algoritme pas a pas, indicant clarament quines operacions farà cada funció.

» Escriure el codi de cada funció.

<br>

**✅ Resultat esperat**
Un DataFrame o diccionari final amb:

- id_cliente
- nombre
- edad
- grupo_edad
- codigo_comportamiento
- segment
- missatge
- factor_codi
- factor_edat
- risc_final


I capacitat per respondre totes les preguntes de màrqueting demanades en una funció.

L'eina seguirà el següent fluxe:
1. Importar pandas per poder treballar amb dataframe.
2. Modificar opcions de la visualització de resultats.
3. Generar els diccionaris constants per poder segmentar.
4. Obtenir el segment i factor de comportament mitjantçant obtenir_segment_factor().
5. Obtenir el segment/factor d'edat mitjançant obtenir_factor_edat().
6. Calcular el risc mitjantçant calcular_risc().
7. Processar els clients mitjançant processar_clients().
8. Generar el dataframe mitjançant crear_dataframe().
9. Respondre les respostes mitjantçant generar_respostes().
10. Carregar les dades dels clients mitjançant la llista de diccionaris clients.
11. Coordinar tot el procés amb main_marketing().
12. Executar l'eina amb main_marketing().

Importa la llibreria pandas.

In [13]:
import pandas as pd

1. Centra els títols de les columnes en la visualització dels resultats.
2. Defineix l'amplada màxima de les linies en la visualització dels resultats.
3. Defineix l'amplada màxima de les columnes en la visualització dels resultats.

In [14]:
pd.set_option("display.colheader_justify", "center")
pd.set_option("display.width", 120)
pd.set_option('display.max_colwidth', None)

Es generen diccionaris constants per poder realitzar la segmentació.

In [15]:
FACTORS_COMPORTAMENT = {
    0: 0.10,
    1: 0.25,
    2: 0.50,
    3: 0.75,
    4: 0.90
}

FACTORS_EDAT = {
    "18-25": 1.20,
    "26-40": 1.00,
    "41-60": 0.85,
    "60+": 0.70
}

SEGMENTS = {
    0: {"segment": "Navega però no afegeix productes",
        "missatge": "Explora els nostres productes i descobreix noves ofertes per a tu."
    },
    1: {"segment": "Afegeix productes però no arriba al checkout",
        "missatge": "Tens productes a la cistella, torna i revisa la teva selecció."
    },
    2: {"segment": "Arriba al checkout però no paga",
        "missatge": "La teva compra està a punt, finalitza-la en pocs clics."
    },
    3: {"segment": "Cistella de compra abandonat fa menys de 24 h",
        "missatge": "Encara tens la teva cistella disponible, no t'oblidis d'emportar-te el que més t'agrada."
    },
    4: {"segment": "Cistella de compra abandonat fa més de 24 h",
        "missatge": "Recupera la teva cistella amb una oferta especial pensada per a tu."
    }
}

#### obtenir_segment_factor():
1. Recupera el codi_comportament de Clients.
2. Recupera el segment i missatge corresponent de SEGMENTS.
3. Recupera el factor corresponent de FACTOR_COMPORTAMENT.
4. Retorna una tupla amb el segment, el missatge i el factor.

In [16]:
def obtenir_segment_factor(client):
    """
    Va a buscar en els diccionaris constants SEGMENTS i FACTORS_COMPORTAMENT.
    Retorna una tupla amb el segment, missatge i factor_comportament.
    """
    codi = client["codi_comportament"]
    info = SEGMENTS[codi]
    factor = FACTORS_COMPORTAMENT[codi]
    return info["segment"], info["missatge"], factor

#### obtenir_factor_edat():
1. Recupera el grup_edat de clients.
2. Retorna el factor_edat corresponent de FACTORS_EDAT.

In [17]:
def obtenir_factor_edat(client):
    """
    Va a buscar el grup_edat de la llista de diccionaris clients.
    Retorna el factor_edat corresponent.
    """
    grup = client["grup_edat"]
    return FACTORS_EDAT[grup]

#### calcular_risc():
1. Rep el factor_comportament i el factor_edat de processar_clients().
2. Retorna el producte dels dos factors.

In [18]:
def calcular_risc(factor_comportament, factor_edat):
    """
    Rep factor_comportament i factor_edat.
    Retorna el producte de les dues dades.
    """
    return factor_comportament * factor_edat

#### processar_clients():
1. Rep llista de diccionaris _clients_.
2. Genera la llista clients_processats.
3. Recull segment, missatge i factor_comportament de obtenir_segment_factor().
4. Recull factor_edat de obtenir_factor_edat().
5. Recull risc_final de calcular_risc().
6. Genera el diccionari nou_client amb les dades del client i les dades rebudes.
7. Afegeix aquest diccionari a la llista clients_processats.
8. Retorna la llista clients_processats.

In [19]:
def processar_clients(clients):
    """
    Recull segment, missatge, factor_comportament, factor_edat i risc_final.
    Retorna llista clients_processats.
    """
    clients_processats = []

    for client in clients:
        segment, missatge, factor_comportament = obtenir_segment_factor(client)
        factor_edat = obtenir_factor_edat(client)
        risc_final = calcular_risc(factor_comportament, factor_edat)

        nou_client = {
            "id_client": client["id_client"],
            "nom": client["nom"],
            "edat": client["edat"],
            "grup_edat": client["grup_edat"],
            "codi_comportament": client["codi_comportament"],
            "segment": segment,
            "missatge": missatge,
            "factor_comportament": factor_comportament,
            "factor_edat": factor_edat,
            "risc_final": risc_final
        }

        clients_processats.append(nou_client)

    return clients_processats

#### crear_dataframe():
1. Rep la llista clients_processats de processar_clients().
2. Genera el dataframe df_risc a partir de la llista.
3. Retorna el dataframe.

In [20]:
def crear_dataframe(clients_processats):
    """
    Rep llista clients_processats.
    Retorna dataframe df_risc.
    """
    df_risc = pd.DataFrame(clients_processats)
    return df_risc

#### generar_respostes():
1. Rep el dataframe df_risc de crear_dataframe().
2. Ordena el dataframe pel valor risc_final descendement.
3. Mostra en pantalla els 3 clients amb el risc_final més alt.
4. Agrupa els clients amb el mateix codi_comportament i per cada grup amb més d'un client els mostra en pantalla amb les diferencies entre edats i riscos.
5. Filtra clients amb risc_final menor de 0.20 i els mostra en pantalla.
6. Selecciona el client de la posició 0 del dataframe ordenat i el mostra en pantalla.

In [21]:
def generar_respostes(df_risc):
    """
    Rep dataframe df_risc.
    Mostra per pantalla les respostes a les preguntes de marketing.
    """
    df_ordenat = df_risc.sort_values("risc_final", ascending=False)

    print("\n1) Clients a contactar primer (risc més alt):")
    top3 = df_ordenat.head(3)[["nom", "risc_final"]]
    print(top3.to_string(index=False))

    print("\n2) Diferències entre clients amb mateix comportament però edats diferents:")
    grups = df_risc.groupby("codi_comportament")

    for codi, grup in grups:
        if len(grup) > 1:
            info = SEGMENTS[codi]
            print(f"\nCodi {codi}: {info['segment']}")
            print(grup[["nom", "grup_edat", "risc_final"]]
                  .sort_values("grup_edat")
                  .to_string(index=False))

    print("\n3) Clients amb risc tan baix que no val la pena invertir-hi:")
    descartats = df_risc.query("risc_final < 0.20")[["nom", "risc_final"]]
    print(descartats.to_string(index=False))

    print("\n4) Client més rendible de recuperar:")
    millor = df_ordenat.iloc[0]
    print(f"- {millor['nom']} (risc: {millor['risc_final']:.2f})\n")

#### clients:
1. Càrrega de la llista de diccionaris _clients_.

In [22]:
clients = [
    {"id_client": 1, "nom": "Laura Serra", "edat": 23, "grup_edat": "18-25", "codi_comportament": 4},
    {"id_client": 2, "nom": "Marc Vidal", "edat": 35, "grup_edat": "26-40", "codi_comportament": 3},
    {"id_client": 3, "nom": "Ana López", "edat": 29, "grup_edat": "26-40", "codi_comportament": 2},
    {"id_client": 4, "nom": "Joan Riera", "edat": 19, "grup_edat": "18-25", "codi_comportament": 2},
    {"id_client": 5, "nom": "Mónica Pérez", "edat": 47, "grup_edat": "41-60", "codi_comportament": 4},
    {"id_client": 6, "nom": "Luis García", "edat": 52, "grup_edat": "41-60", "codi_comportament": 1},
    {"id_client": 7, "nom": "Pilar Sánchez", "edat": 61, "grup_edat": "60+", "codi_comportament": 3},
    {"id_client": 8, "nom": "Eva Martín", "edat": 38, "grup_edat": "26-40", "codi_comportament": 0},
    {"id_client": 9, "nom": "Diego Romero", "edat": 24, "grup_edat": "18-25", "codi_comportament": 1},
    {"id_client": 10, "nom": "Núria Costa", "edat": 33, "grup_edat": "26-40", "codi_comportament": 4}]

#### main_marketing():
1. Crida la funció processar_clients() i recull la llista clients_processats.
2. Crida la funció crear_dataframe() i genera el dataframe df_risc.
3. Executa generar_respostes().
4. Retorna el dataframe df_risc.

In [23]:
def main_marketing(clients):
    """
    Ordena tot el procés per respondre a les preguntes de marketing.
    """
    clients_processats = processar_clients(clients)
    df_risc = crear_dataframe(clients_processats)
    generar_respostes(df_risc)
    return df_risc


#### main_marketing():
1. Crida la funció main_marketing().

In [24]:
df_risc = main_marketing(clients)
df_risc


1) Clients a contactar primer (risc més alt):
    nom       risc_final
 Laura Serra    1.080   
 Núria Costa    0.900   
Mónica Pérez    0.765   

2) Diferències entre clients amb mateix comportament però edats diferents:

Codi 1: Afegeix productes però no arriba al checkout
    nom      grup_edat  risc_final
Diego Romero   18-25      0.3000  
 Luis García   41-60      0.2125  

Codi 2: Arriba al checkout però no paga
   nom     grup_edat  risc_final
Joan Riera   18-25       0.6    
 Ana López   26-40       0.5    

Codi 3: Cistella de compra abandonat fa menys de 24 h
     nom      grup_edat  risc_final
   Marc Vidal   26-40      0.750   
Pilar Sánchez     60+      0.525   

Codi 4: Cistella de compra abandonat fa més de 24 h
    nom      grup_edat  risc_final
 Laura Serra   18-25      1.080   
 Núria Costa   26-40      0.900   
Mónica Pérez   41-60      0.765   

3) Clients amb risc tan baix que no val la pena invertir-hi:
   nom      risc_final
Eva Martín     0.1    

4) Client més

,id_client,nom,edat,grup_edat,codi_comportament,segment,missatge,factor_comportament,factor_edat,risc_final
0,1,Laura Serra,23,18-25,4,Cistella de compra abandonat fa més de 24 h,Recupera la teva cistella amb una oferta especial pensada per a tu.,0.90,1.20,1.0800
1,2,Marc Vidal,35,26-40,3,Cistella de compra abandonat fa menys de 24 h,"Encara tens la teva cistella disponible, no t'oblidis d'emportar-te el que més t'agrada.",0.75,1.00,0.7500
2,3,Ana López,29,26-40,2,Arriba al checkout però no paga,"La teva compra està a punt, finalitza-la en pocs clics.",0.50,1.00,0.5000
3,4,Joan Riera,19,18-25,2,Arriba al checkout però no paga,"La teva compra està a punt, finalitza-la en pocs clics.",0.50,1.20,0.6000
4,5,Mónica Pérez,47,41-60,4,Cistella de compra abandonat fa més de 24 h,Recupera la teva cistella amb una oferta especial pensada per a tu.,0.90,0.85,0.7650
5,6,Luis García,52,41-60,1,Afegeix productes però no arriba al checkout,"Tens productes a la cistella, torna i revisa la teva selecció.",0.25,0.85,0.2125
6,7,Pilar Sánchez,61,60+,3,Cistella de compra abandonat fa menys de 24 h,"Encara tens la teva cistella disponible, no t'oblidis d'emportar-te el que més t'agrada.",0.75,0.70,0.5250
7,8,Eva Martín,38,26-40,0,Navega però no afegeix productes,Explora els nostres productes i descobreix noves ofertes per a tu.,0.10,1.00,0.1000
8,9,Diego Romero,24,18-25,1,Afegeix productes però no arriba al checkout,"Tens productes a la cistella, torna i revisa la teva selecció.",0.25,1.20,0.3000
9,10,Núria Costa,33,26-40,4,Cistella de compra abandonat fa més de 24 h,Recupera la teva cistella amb una oferta especial pensada per a tu.,0.90,1.00,0.9000


### Comentari sobre els resultats obtinguts

Els resultats mostren que la combinació entre comportament i grup d’edat permet prioritzar de manera clara quins clients s’han de contactar primer en una campanya de recuperació de cistelles abandonades. Els perfils més joves amb cistella abandonada fa més de 24 hores presenten el risc més alt, mentre que els clients de més edat, tot i tenir el mateix comportament, mostren un risc inferior. També s’identifiquen perfils amb interès molt baix, com el cas d’Eva Martín, on no és eficient invertir-hi pressupost. En canvi, clientes com Laura Serra representen oportunitats de recuperació amb un alt potencial de conversió i, per tant, són les més rendibles de prioritzar.

---
---

# Nivell 2

## Exercici 1: Gestió de risc i priorització de pacients

Treballes com a analista de dades en un centre mèdic especialitzat. Diversos pacients han enviat els seus símptomes a través d'un formulari online, però les dades arriben en un format desordenat i poc estructurat, i l'equip mèdic necessita que els organitzis per poder prioritzar l'ordre d'atenció segons la gravetat.

A continuació tens una llista amb informació real procedent del sistema:

```
pacients = [
    "Maria|45|tos, Fiebre Alta, dolor_pecho",
    "Luis|33|dolor_cabeza,faTiga",
    "Sara|67|fiebre alta, dificultad RESPIRAR, tos",
    "Jordi|52|fatiga , tos",
    "Anna|29|fiebre alta,dolor_cabeza"]
```
                
El departament mèdic també t'ha proporcionat el següent diccionari de nivells de risc per símptoma:
                  
```
nivells_risc = {
    "fiebre_alta": 3,
    "dificultad_respirar": 5,
    "dolor_pecho": 4,
    "tos": 1,
    "fatiga": 1,
    "dolor_cabeza": 1}
```

<br>

**🎯 Objectius**
L'objectiu és transformar aquestes dades en informació útil per als metges. A partir dels registres en brut, has de:

» Convertir cada string en una estructura neta (diccionari o similar).

» Separar i normalitzar la llista de símptomes:
- minúscules
- eliminar espais
- dividir correctament per comes

» Assignar un risc a cada símptoma utilitzant el diccionari nivells_risc.

» Calcular el risc total de cada pacient sumant els valors.

» Ordenar els pacients de major a menor risc per determinar la prioritat d'atenció.

<br>

**⚠️ Important**

En aquest Nivell 2 ja no et proporcionem la llista de funcions que has de crear. Ets tu qui ha de plantejar quines funcions necessites i com modularitzar el procés.

Tot i així, es recomana que:

    » La funció principal orquestri tots els passos.

    » Cada subfunció faci una sola tasca concreta (neteja, conversió, càlcul, ordenació...).

    » La lògica estigui separada de manera neta i coherent.

<br>

**✅ Resultat esperat**
Un DataFrame o un diccionari amb:

- nom
- edat
- símptomes normalitzats
- risc_total

I un apartat final amb els pacients ordenats per prioritat (més risc → primer).

Es seguirà el següent fluxe:
1. Importar pandas per poder treballar amb dataframe.
2. Modificar opcions de la visualització de resultats.
3. Separar els registres per poder-los netejar mitjançant la funció separar_registres().
4. Separar els símptomes per poder-los netejar mitjançant la funció separar_simptomes().
5. Definir la funció netejar_simptoma() per la posterior neteja de simptomes.
6. Netejar els símptomes mitjançat la funció netejar_llista_simptomes().
7. Calcular el risc per pacient mitjançant la funció calcular_risc_pacient().
8. Crear diccionari per pacient mitjançant la funció crear_diccionar_pacient().
9. Crear dataframe mitjançant la funció generar_dataframe().
10. Ordenar els pacients mitjançant la funció ordenar_pacients().
11. Carregar les dades dels pacients.
12. Carregar els nivells de risc.
13. Coordinar tot el procés mitjantçant la funció main_gestio_risc().
14. Executar tot el procés amb main_gestio_risc().

Importa la llibreria pandas.

In [25]:
import pandas as pd

Defineix l'amplada màxima de les columnes en la visualització dels resultats.

In [26]:
pd.set_option('display.max_colwidth', None)

##### separar_registres():
1. Rep el registre de la llista _pacients_.
2. Separa per nom, edat i símptomes.
3. Retorna nom, edat tipus enter i símptomes.

In [27]:
def separar_registre(registre):
    """
    Rep el registre de pacients.
    Retorna separats nom, edat i símptomes.
    """
    nom, edat, simptomes = registre.split("|")
    return nom, int(edat), simptomes

#### separar_simptomes():
1. Rep string de simptomes.
2. Separa simptomes per ",".
3. Retorna simptomes separats.

In [28]:
def separar_simptomes(simptomes_str):
    """
    Rep string de símptomes.
    Retorna símptomes separats per ",".
    """
    return simptomes_str.split(",")

#### netejar_simptoma():
1. Rep símptoma de netejar_llista_simptomes().
2. Neteja espais en blanc al principi i al final de la paraula.
3. Converteix tots els caràcters a minúscules.
4. Substitueix espais en blanc interns per guions baixos.
5. Retorna el simptoma net.


In [29]:
def netejar_simptoma(simptoma):
    """
    Rep simptomes_bruts.
    Retorna simptomes sense espais en blanc, en mínuscules i guions baixos (_).
    """
    return simptoma.strip().lower().replace(" ", "_")

#### netejar_llista_simptomes():
1. Rep llista de símptomes bruts de main_gestio_risc().
2. Executa netejar_simptoma() per cada simptoma de la llista.
3. Retorna llista de símptomes nets.

In [30]:
def netejar_llista_simptomes(llista):
    """
    Rep llista simptomes_bruts.
    Executa netejar_simptoma() per cada símptoma i retorna llista simptomes_nets.
    """
    return [netejar_simptoma(simptoma) for simptoma in llista]

#### calcular_risc_pacient():
1. Rep simptomes i el diccionari nivells_risc de main_gestio_risc().
2. Busca per cada simptoma el seu nivell de risc.
3. Si no troba el nivell de risc, li assigna 0.
4. Retorna el sumatori de tots els nivells de risc.

In [31]:
def calcular_risc_pacient(simptomes, nivells_risc):
    """
    Rep llista simptomes_nets i diccionari nivells_risc.
    Busca valors corresponents per simptoma i retorna sumatori.
    """
    return sum(nivells_risc.get(simptoma, 0) for simptoma in simptomes)

#### crear_diccionari_pacient():
1. Rep nom, edat, simptomes i risc_total de main_gestio_risc().
2. Retorna un diccionari amb aquestes dades.

In [32]:
def crear_diccionari_pacient(nom, edat, simptomes, risc_total):
    """
    Rep nom, edat, simptomes i risc_total.
    Retorna diccionari.
    """
    return {
        "nom": nom,
        "edat": edat,
        "simptomes": simptomes,
        "risc_total": risc_total
    }

#### generar_dataframe()_
1. Rep llista de diccionaris llista_resultats.
2. Converteix llista de diccionaris a dataframe df_gestio_risc.
3. Retorna el dataframe generat.

In [33]:
def generar_dataframe(llista_diccs):
    """
    Rep llista de diccionaris llista_resultats.
    Retorna dataframe df_gestio_risc.
    """
    df_gestio_risc = pd.DataFrame(llista_diccs)
    return df_gestio_risc

#### ordenar_pacients():
1. Rep dataframe de main_gestio_risc().
2. Ordena el dataframe pel valor risc_total descendentment.
3. Retorna el dataframe ordenat.

In [34]:
def ordenar_pacients(df):
    """
    Rep dataframe df_gestio_risc.
    Retorna dataframe df_gestio_risc ordenat.
    """
    return df.sort_values(by="risc_total", ascending=False)

#### pacients:
1. Càrrega de la llista de cadenes de text _pacients_.

In [35]:
pacients = [
    "Maria|45|tos, Fiebre Alta, dolor_pecho",
    "Luis|33|dolor_cabeza,faTiga",
    "Sara|67|fiebre alta, dificultad RESPIRAR, tos",
    "Jordi|52|fatiga , tos",
    "Anna|29|fiebre alta,dolor_cabeza"
]

#### nivells_risc:
1. Càrrega del diccionari _nivells_risc_.

In [36]:
nivells_risc = {
    "fiebre_alta": 3,
    "dificultad_respirar": 5,
    "dolor_pecho": 4,
    "tos": 1,
    "fatiga": 1,
    "dolor_cabeza": 1
}

#### main_gestio_risc():
1. Rep la llista _pacients_ i el diccionari _nivells_risc_.
2. Genera llista_resultats.
3. Crida la funció separar_registre() i rep nom, edat_simtomes_str.
4. Crida la funció separar_simptomes() i rep simptomes_bruts.
5. Crida la funció netejar_llista_simptomes() i rep simptomes_nets.
6. Crida la funció calcular_risc_pacient() i rep risc_total.
7. Crida la funció crear_diccionari_pacient() i rep pacient_dicc.
8. Afegeix diccionari pacient_dic a llista_resultats.
9. Crida la funció generar_dataframe() i rep df_gestio_risc.
10. Crida la funció ordenar_pacients() i rep dataframe ordenat.
11. Retorna el dataframe df_gestio_risc.

In [37]:
def main_gestio_risc(pacients, nivells_risc):
    """
    Ordena i executa les funcions per generar el dataframe.
    Retorna el dataframe final amb la gestió de risc i priorització de pacients.
    """
    llista_resultats = []

    for registre in pacients:
        nom, edat, simptomes_str = separar_registre(registre)
        simptomes_bruts = separar_simptomes(simptomes_str)
        simptomes_nets = netejar_llista_simptomes(simptomes_bruts)
        risc_total = calcular_risc_pacient(simptomes_nets, nivells_risc)

        pacient_dicc = crear_diccionari_pacient(nom, edat, simptomes_nets, risc_total)

        llista_resultats.append(pacient_dicc)

    df_gestio_risc = generar_dataframe(llista_resultats)
    return ordenar_pacients(df_gestio_risc)

#### main_gestio_risc():
1. Crida la funció main_gestio_risc().

In [38]:
df_gestio_risc = main_gestio_risc(pacients, nivells_risc)
df_gestio_risc

,nom,edat,simptomes,risc_total
2,Sara,67,"[fiebre_alta, dificultad_respirar, tos]",9
0,Maria,45,"[tos, fiebre_alta, dolor_pecho]",8
4,Anna,29,"[fiebre_alta, dolor_cabeza]",4
1,Luis,33,"[dolor_cabeza, fatiga]",2
3,Jordi,52,"[fatiga, tos]",2


### Comentari sobre els resultats obtinguts
L’anàlisi permet identificar de manera clara la prioritat d’atenció mèdica segons el risc total calculat a partir dels símptomes. Els pacients Sara i Maria presenten els valors més elevats, associats a símptomes de major gravetat com dificultat per respirar o dolor al pit, que requereixen una resposta immediata. En canvi, Luis i Jordi mostren un risc molt baix, indicant que poden ser atesos en una segona fase. Aquesta segmentació facilita una gestió eficient dels recursos mèdics, assegurant que els casos més urgents rebin atenció prioritària i que el flux de pacients es mantingui ordenat segons la gravetat clínica.

---
---

## Exercici 2: Classificació de clients de segurs

Treballes com a analista de dades en una companyia asseguradora.
L'empresa vol calcular el risc dels clients, el preu estimat de la pòlissa i la probabilitat de frau, utilitzant un codi de comportament que ha estat generat per un model intern.

Tens la següent llista de clients:

```
clients = [
    {"nom": "Ana", "edat": 42, "codi": 0},
    {"nom": "Carlos", "edat": 30, "codi": 2},
    {"nom": "Isabel", "edat": 55, "codi": 3},
    {"nom": "Jorge", "edat": 40, "codi": 1},
    {"nom": "Marta", "edat": 28, "codi": 0}]
```
I aquest diccionari amb les categories del sistema:

```
nivells_seguro = {
    0: {"categoria": "Baix risc", "preu_base": 120, "fraude": 0.01},
    1: {"categoria": "Risc mitjà", "preu_base": 200, "fraude": 0.05},
    2: {"categoria": "Risc alt", "preu_base": 350, "fraude": 0.15},
    3: {"categoria": "Risc crític", "preu_base": 500, "fraude": 0.30}}
```

El preu final s'ha de calcular amb la fórmula:

```preu_final = preu_base * (1 + edat/100)```

<br>

**Objectius**

» Classificar cada client utilitzant nivells_seguro (sense if/elif).

» Afegir:
- La categoria
- El preu base
- La probabilitat de frau

» Calcular el preu final de la pòlissa amb la fórmula donada.

» Crear un DataFrame o diccionari amb la informació final.

» Ordenar els clients:
- Per probabilitat de frau (descendent).
- Per preu final.

<br>

 **Important**

Igual que a l'Exercici 1, en aquest nivell ets tu qui ha de definir les funcions necessàries. Es recomana mantenir el codi modular, on:

» Cada funció resol un pas del procés.

» La funció principal delega totes les tasques.

<br>

**✅ Resultat esperat**
Un DataFrame o un diccionari amb:

- Nom
- Edat
- Categoria
- Preu_base
- Preu_final
- Probabilitat_frau
- Un apartat d'anàlisi on s'indiqui:
    - Quin client té més probabilitat de frau
    - Quin té el preu final més alt
    - Qui seria prioritari de revisar manualment


Es seguirà el següent fluxe:
1. Importar pandas per poder treballar amb dataframe.
2. Modificar opcions de la visualització de resultats.
3. Classificar els clients mitjançant la funció classificar_client().
4. Calcular el preu de la polissa mitjançant la funció calcular_preu_polissa().
5. Processar les dades del client mitjançant la funció processar_clients_seguro().
6. Crear el dataframe mitjançant la funció crear_dataframe_seguro().
7. Respondre a l'anàlisi mitjançant la funció generar_analisi().
8. Carregar les dades dels clients.
9. Carregar els nivells de risc de l'asseguradora.
10. Coordinar tot el procés mitjantçant la funció main_seguro().
11. Executar tot el procés amb main_seguro().

Importa la llibreria pandas.

In [39]:
import pandas as pd

Defineix l'amplada màxima de les columnes en la visualització dels resultats.

In [40]:
pd.set_option('display.max_colwidth', None)

#### classificar_client():
1. Rep les dades de processar_clients_seguro().
2. Genera el diccionari info a partir del diccionari nivells_seguro.
3. Retorna diccionari amb les dades de client i info.

In [41]:
def classificar_client(client, nivells_seguro):
    """
    Rep dades i llista de diccionaris.
    Retorna diccionari.
    """
    info = nivells_seguro[client["codi"]]
    return {
        "nom": client["nom"],
        "edat": client["edat"],
        "categoria": info["categoria"],
        "preu_base": info["preu_base"],
        "prob_frau": info["fraude"]
    }

#### calcular_preu_polissa():
1. Rep les dades de processar_clients_seguro().
2. Calcula el preu de la pòlissa a partir d'una formula estipulada anteriorment.
3. Retorna el preu de la pòlissa.

In [42]:
def calcular_preu_polissa(preu_base, edat):
    """
    Rep dades de preu i edat.
    Retorna cálcul del preu de la pòlissa.
    """
    return preu_base * (1 + edat/100)

#### processar_clients_seguro():
1. Genera la llista de diccionaris resultat.
2. Per cada client
    - Crida classificar_client() i recupera diccionari.
    - Crida calcular_preu_polissa i recupera preu_final.
3. Afegeix les dades recuperades a la llista resultat.
4. Retorna la llista de diccionaris resultat.


In [43]:
def processar_clients_seguro(clients_seguro, nivells_seguro):
    """
    Rep les dades de clients i nivells de risc.
    Retorna llista de diccionaris.
    """
    resultat = []
    for client in clients_seguro:
        dades = classificar_client(client, nivells_seguro)
        dades["preu_final"] = calcular_preu_polissa(dades["preu_base"], dades["edat"])
        resultat.append(dades)
    return resultat

#### crear_dataframe_seguro():
1. Rep llista de diccionaris resultat.
2. Converteix llista de diccionaris a dataframe df_seguro.
3. Retorna el dataframe df_seguro.

In [44]:
def crear_dataframe_seguro(llista):
    """
    Rep llista de diccionaris.
    Retorna dataframe.
    """
    df_seguro = pd.DataFrame(llista)
    return df_seguro

#### generar_analisi():
1. Rep de main_seguro el dataframe df_seguro.
2. Ordena df_seguro pel valor prob_frau i mostra la primera posició per pantalla.
3. Ordena df_seguro pel valor preu_final i mostra la primera posició per pantalla.
4. Mostra per pantalla la primera posició top_frau com a revisió prioritària.

In [45]:
def generar_analisi(df_seguro):
    """
    Rep dataframe.
    Mostra per pantalla el client amb més probabilitat de frau, el que té el preu de la pòlissa més alt i el que s'ha de revisar prioritariament.
    """
    print("\n Client amb més probabilitat de frau:")
    top_frau = df_seguro.sort_values("prob_frau", ascending=False).iloc[0]
    print(f"- {top_frau['nom']} ({top_frau['prob_frau']:.2f})")

    print("\n Client amb el preu final més alt:")
    top_preu = df_seguro.sort_values("preu_final", ascending=False).iloc[0]
    print(f"- {top_preu['nom']} ({top_preu['preu_final']:.2f} €)")

    print("\n Prioritari de revisar manualment:")
    print(f"- {top_frau['nom']} (risc de frau més alt)\n")

#### clients_seguro:
1. Carrega la llista de diccionars _clients_seguro_.

In [46]:
clients_seguro = [
    {"nom": "Ana", "edat": 42, "codi": 0},
    {"nom": "Carlos", "edat": 30, "codi": 2},
    {"nom": "Isabel", "edat": 55, "codi": 3},
    {"nom": "Jorge", "edat": 40, "codi": 1},
    {"nom": "Marta", "edat": 28, "codi": 0}]

#### nivells_seguro:
1. Carrega el diccionari de diccionars _nivells_seguro_.

In [47]:
nivells_seguro = {
    0: {"categoria": "Baix risc", "preu_base": 120, "fraude": 0.01},
    1: {"categoria": "Risc mitjà", "preu_base": 200, "fraude": 0.05},
    2: {"categoria": "Risc alt", "preu_base": 350, "fraude": 0.15},
    3: {"categoria": "Risc crític", "preu_base": 500, "fraude": 0.30}}

#### main_seguro():
1. Crida la funció processar_clients_seguro() i rep processats.
2. Crida la funció crear_dataframe_seguro() i rep df_seguro.
3. Crida la funció generar_analisi().
4. Ordena el dataframe df_seguro.
4. Retorna el dataframe df_seguro_ordenat.

In [48]:
def main_seguro(clients_seguro, nivells_seguro):
    """
    Ordena tot el procés per respondre als objetius de l'empresa.
    """
    processats = processar_clients_seguro(clients_seguro, nivells_seguro)
    df_seguro = crear_dataframe_seguro(processats)
    generar_analisi(df_seguro)
    df_seguro_ordenat = df_seguro.sort_values(by = ["prob_frau", "preu_final"], ascending = [False,False])
    return df_seguro_ordenat

#### main_seguro():
1. Crida la funció main_seguro().

In [49]:
df_seguro_ordenat = main_seguro(clients_seguro, nivells_seguro)
df_seguro_ordenat


 Client amb més probabilitat de frau:
- Isabel (0.30)

 Client amb el preu final més alt:
- Isabel (775.00 €)

 Prioritari de revisar manualment:
- Isabel (risc de frau més alt)



,nom,edat,categoria,preu_base,prob_frau,preu_final
2,Isabel,55,Risc crític,500,0.30,775.0
1,Carlos,30,Risc alt,350,0.15,455.0
3,Jorge,40,Risc mitjà,200,0.05,280.0
0,Ana,42,Baix risc,120,0.01,170.4
4,Marta,28,Baix risc,120,0.01,153.6


### Comentari sobre els resultats obtinguts
L’anàlisi permet classificar els clients segons el seu nivell de risc i calcular el preu final de la pòlissa en funció de l’edat. Els resultats mostren que Isabel presenta el risc més alt i la probabilitat de frau més elevada, fet que la situa com a client prioritari per a revisió manual. Els perfils de Carlos i Jorge també indiquen nivells de risc significatius, però amb menor impacte econòmic. En canvi, Ana i Marta pertanyen al segment de baix risc, amb preus finals més baixos i una probabilitat de frau mínima. Aquesta classificació facilita la presa de decisions en la gestió de pòlisses, permetent optimitzar la supervisió de casos crítics i ajustar estratègies de preus segons el perfil de cada client.

---
---

# Nivell 3

## Exercici 1: Generador de contrasenyes segures

Ets analista de dades dins el departament de Ciberseguretat d'una empresa tecnològica. S'ha detectat que la majoria de treballadors continuen utilitzant contrasenyes molt dèbils com “Asdf1234”, dates d'aniversari o combinacions fàcils d'endevinar. Per aquest motiu, el responsable del departament t'ha demanat desenvolupar un generador de contrasenyes segures, totalment parametritzable i modular.

<br>

**📝 Requisits del generador**
Has de crear una funció principal i totes les subfuncions necessàries per:

» Generar contrasenyes de longitud variable.

» Decidir si han d'incloure:
- Majúscules
- Minúscules
- Números
- Signes especials
» Assegurar-te que com a mínim apareix un caràcter de cada tipus seleccionat.

» Aleatoritzar completament la contrasenya.

» Retornar la contrasenya generada.

<br>

```
📍 Nota important
Explora el funcionament del mòdul random de la llibreria numpy
```
<br>

La funció principal del generador ha de tenir valors per defecte, exactament els següents:

- majuscules=True
- minuscules=True
- numeros=True
- signes=False

Aquests paràmetres són obligatoris i formen part del disseny del sistema.

**👁️ Exemple d'ús esperat**

```crear_contrasenya(12,majuscules=True,minuscules=True,numeros=True,signes=True)```

Output possible:

```WS6MwfRD:%B3```

Es seguirà el següent fluxe:
1. Importar numpy i string per poder treballar amb el generador random i recopilar caràcters.
2. Recopilar caràcters (majúscules, minúscules, nombres i signes) mitjançant la funció recopilar_caracters().
3. Definir la quantitat de caràcters aleatoris mínims indicats pels paràmetres mitjançant la funció definir_minims().
4. Definir la quantitat de caràcters aleatoris restants mitjançant la funció definir_restants().
5. Remenar els caràcters mínims i restants per generar la contrasenya mitjançant la funció remenar_caracters().
6. Coordinar tot el procés mitjançant la funció main_contrasenya().
7. Executar tot el procés amb main_contrasenya().

Importa la llibreria numpy i el mòdul string.

In [50]:
import numpy as np
import string

#### recopilar_caracters():
1. Rep els paràmetres booleans de main_contrasenya().
2. Recopila les dades corresponents als paràmetres que son True.
3. Retorna diccionari.

In [51]:
def recopilar_caracters(majuscules, minuscules, numeros, signes):
    """
    Rep els paràmetres booleans.
    Retorna diccionari.
    """
    conjunts = {}

    if majuscules:
        conjunts["majuscules"] = list(string.ascii_uppercase)
    if minuscules:
        conjunts["minuscules"] = list(string.ascii_lowercase)
    if numeros:
        conjunts["numeros"] = list(string.digits)
    if signes:
        conjunts["signes"] = list(string.punctuation)

    return conjunts

#### definir_minims():
1. Rep el generador aleatori i el diccionari conjunts.
2. Genera la llista minims.
3. Selecciona un caracter aleatori de cada conjunt.
4. Afegeix el caràcter a la llista minims.
5. Retorna la llista minims.

In [52]:
def definir_minims(aleatori, conjunts):
    """
    Rep el generador aleatori i un diccionari.
    Retorna llista.
    """
    minims = []
    for llista in conjunts.values():
        minims.append(aleatori.choice(llista))
    return minims

#### definir_restants():
1. Rep el generador aleatori, el diccionari conjunts i la quantitat de caràcters restants.
2. Genera la llista tots.
3. Omple la llista amb tots els caràcters disponibles.
4. Selecciona aleatoriaments tants caràcters com la quantitat de caràcters restants rebuda.
5. Retorna la llista amb els caràcters restants seleccionats.

In [53]:
def definir_restants(aleatori, conjunts, longitud_rest):
    """
    Rep el generador aleatori, un diccionari i una quantitat.
    Retorna llista.
    """
    tots = []
    for llista in conjunts.values():
        tots.extend(llista)
    return list(aleatori.choice(tots, size=longitud_rest))

#### remenar_caracters():
1. Rep el generador aleatori i una llista amb els caracters (minims + restants).
2. Desordena els caràcters de la llista.
3. Retorna la llista sense espais i convertida a string.

In [54]:
def remenar_caracters(aleatori, llista):
    """
    Rep el generador aleatori i una llista.
    Retorna llista convertida a string sense espais.
    """
    aleatori.shuffle(llista)
    return "".join(llista)

#### main_contrasenya():
1. Rep els paràmetres longitud, majuscules, minuscules, numeros i signes.
2. Crea el generador aleatori amb la funció np.random.default_rng().
3. Crida la funció recopilar_caracters() i recupera diccionari conjunts.
4. Comproba si conjunts està buit i en aquest cas mostra avís per pantalla.
5. Crida la funció definir_minims() i recupera la llista minims.
6. Crida la funció definir_restants() i recupera llista restants.
7. Genera la llista contrasenya concatenant les llistes minims i restants.
8. Crida la funció remenar_caracters() i rep un string amb la contrasenya final.
9. Retorna la contrasenya.

In [55]:
def main_contrasenya(longitud=12, majuscules=True, minuscules=True, numeros=True, signes=False):
    """
    Ordena tot el procés per generar contrasenya aleatòria.
    Retorna contrasenya aleatòria.
    """

    aleatori = np.random.default_rng()

    conjunts = recopilar_caracters(majuscules, minuscules, numeros, signes)

    if not conjunts:
        raise ValueError("Has de seleccionar com a mínim un tipus de caràcter.")

    minims = definir_minims(aleatori, conjunts)

    restants = definir_restants(aleatori, conjunts, longitud - len(minims))

    contrasenya = minims + restants

    return remenar_caracters(aleatori, contrasenya)

#### main_contrasenya():
1. Crida la funció main_contrasenya().

In [56]:
main_contrasenya(12, True, True,True,True)

'1/Pq{5[W>1e/'

### Comentari final
En aquest exercici no té sentit analitzar cada contrasenya generada de forma individual, ja que el resultat és aleatori per disseny. El valor real rau en el fet que l’eina permet definir paràmetres de seguretat (tipus de caràcters i longitud) i garanteix que les contrasenyes compleixin uns criteris mínims de robustesa. Això contribueix a reduir l’ús de contrasenyes febles entre els treballadors i a millorar el nivell global de seguretat de l’organització.

---
---

## Exercici 2: Processament automàtic de dades esportives


Ara formes part d'un equip d'analítica dins una organització esportiva que gestiona dades de la lliga catalana. La teva companya de feina vol automatitzar la lectura del seu arxiu històric de resultats i extreure les estadístiques més importants. T'ha demanat fer un petit sistema en Python que processi el fitxer 📄 "historic partits.txt"

<br>

**🎯 Objectius**

Mitjançant un codi modular, clar i documentat, el teu programa ha de calcular:

- 1.Quin equip és el més golejador.

- 2.Quin és el més golejat.

- 3.El nombre total de gols que ha fet cada equip.

- 4.La classificació final ordenada per punts on cada victòria suma 3 pts, cada empat 1 pts y cada derrota 0 pts.


En aquest nivell no es dona cap funció. Ets tu qui ha de dissenyar un flux complet. El codi ha de reflectir una bona organització interna.

<br>

**✅ Resultat esperat**

Un DataFrame o diccionari final semblant a:

imatge taula equips


A més del text:

Equip més golejador: __

Equip més golejat: __

Classificació final:
1.  __
2.  __
...

L'eina seguirà el següent fluxe:
1. Importar pandas per poder treballar amb dataframe.
2. Carregar l'arxiu historic_partits.txt
3. Netejar les dades mitjançant netejar_linia().
4. Desglossar las linies mitjançant la funció desglossar_linia().
5. Processar els partits mitjançant la funció llegir_partits().
6. Convertir la llista a dataframe mitjançant la funció construir_dataframe().
7. Crear cada equip amb les seves corresponents dades al diccionari equips mitjançant la funció crear_equip().
8. Actualitzar els gols marcats i rebuts mitjançant la funció actualitzar_gols().
9. Actualitzar els resultats dels partits mitjançant la funció actualitzar_resultat().
10. Aplicar al diccionari les actualitzacions de gols, punts i resultats mitjançant la funció processar_partit().
11. Aplicar processar_partits()a tots els partits mitjançant la funció actualitzar_equips().
12. Generar la classificació mitjançant la funció generar_classificacio().
13. Coordinar tot el procés mitjançant la funció main_futbol().
14. Executar tot el procés amb main_futbol().


Importa la llibreria pandas.

In [57]:
import pandas as pd

Carrega les dades a analitzar.

In [58]:
from google.colab import files
uploaded = files.upload()

Saving historic_partits.txt to historic_partits.txt


#### netejar_linia():
1. Rep una linia equivalent a un partit.
2. Treu els espais sobrants.
3. Comproba que no estigui buida.
4. Ho separa en paraules.
5. Retorna la linia.

In [59]:
def netejar_linia(linia):
    """
    Rep una linia i la neteja.
    Retorna linia netejada.
    """
    linia = linia.strip()
    if not linia:
        return None
    return linia.split()

#### desglossar_partit():
1. Rep les paraules netes.
2. Busca la posicio del marcador.
3. Separa el nom de l'equip local i visitant.
4. Extreu els gols del marcador.
5. Retorna un diccionari.

In [60]:
def desglossar_partit(parts):
    """
    Rep les paraules netes.
    Retorna un diccionari.
    """
    posicio_marcador = next(posicio for posicio, paraula in enumerate(parts) if "-" in paraula)

    equip_local = " ".join(parts[:posicio_marcador])
    marcador = parts[posicio_marcador]
    equip_visitant = " ".join(parts[posicio_marcador+1:])

    gols_local, gols_visitant = map(int, marcador.split("-"))

    return {
        "local": equip_local,
        "visitant": equip_visitant,
        "gols_local": gols_local,
        "gols_visitant": gols_visitant
    }

#### llegir_partits():
1. Rep el nom del fitxer a llegir.
2. Crea la llista partits.
3. Envía cada linia a netejar_linia().
4. Envía la linia neta a desglossar_partit().
5. Afegeix els partits a la llista.
6. Retorna la llista partits.

In [61]:
def llegir_partits(nom_fitxer):
    """
    Rep el nom del fitxer a llegir.
    Retorna la llista partits.
    """
    partits = []

    with open(nom_fitxer, "r", encoding="utf-8") as f:
        for linia in f:
            parts = netejar_linia(linia)
            if parts is None:
                continue

            partit = desglossar_partit(parts)
            partits.append(partit)

    return partits

#### construir_dataframe():
1. Rep la llista partits.
2. Converteix la llista al dataframe df_futbol.
3. Retorna el dataframe df_futbol.

In [62]:
def construir_dataframe(partits):
    """
    Rep la llista partits.
    Retorna el dataframe df_futbol.
    """
    df_futbol = pd.DataFrame(partits)
    return df_futbol

#### crear_equip():
1. Rep el diccionari equips i un nom.
2. Comprova si l'equip existeix en el diccionari.
3. Si no existeix, crea l'equip al diccionari amb les estadístiques a 0.

In [63]:
def crear_equip(equips, nom):
    """
    Rep el diccionari equips i un nom.
    Crea l'equip al diccionari amb les estadístiques a 0.
    """
    if nom not in equips:
        equips[nom] = {
            "gf": 0,
            "gc": 0,
            "guanyats": 0,
            "empatats": 0,
            "perduts": 0,
            "punts": 0
        }

#### actualitzar_gols():
1. Rep el diccionari equips i les dades del partit.
2. Suma els gols a favor i en contra de l'equip local.
3. Suma els gols a favor i en contra de l'equip visitant.

In [64]:
def actualitzar_gols(equips, local, visitant, gols_local, gols_visitant):
    """
    Rep el diccionari equips i les dades del partit.
    Suma els gols a favor i en contra de l'equip local.
    Suma els gols a favor i en contra de l'equip visitant.
    """
    equips[local]["gf"] += gols_local
    equips[local]["gc"] += gols_visitant
    equips[visitant]["gf"] += gols_visitant
    equips[visitant]["gc"] += gols_local

#### actualitzar_resultat():
1. Rep el diccionari equips i les dades del partit.
2. Determina si ha guanyat l'equip local, el visitant o si ha estat empat.
3. Assigna els punts corresponents.

In [65]:
def actualitzar_resultat(equips, local, visitant, gols_local, gols_visitant):
    """
    Rep el diccionari equips i les dades del partit.
    Assigna els punts corresponents per victoria, derrota o empat.
    """
    if gols_local > gols_visitant:
        equips[local]["guanyats"] += 1
        equips[visitant]["perduts"] += 1
        equips[local]["punts"] += 3

    elif gols_visitant > gols_local:
        equips[visitant]["guanyats"] += 1
        equips[local]["perduts"] += 1
        equips[visitant]["punts"] += 3

    else:
        equips[local]["empatats"] += 1
        equips[visitant]["empatats"] += 1
        equips[local]["punts"] += 1
        equips[visitant]["punts"] += 1

#### processar_partits():
1. Rep el diccionari equips i les dades del partit.
2. Crida crear_equip() per crear els equips.
3. Crida actualitzar_gols() per actualitzar els gols.
4. Crida actualitzar_resultat() per actualitzar els resultats.

In [66]:
def processar_partit(equips, local, visitant, gols_local, gols_visitant):
    """
    Rep el diccionari equips i les dades del partit.
    Crea els equips si cal i actualitza gols i resultats.
    """
    crear_equip(equips, local)
    crear_equip(equips, visitant)
    actualitzar_gols(equips, local, visitant, gols_local, gols_visitant)
    actualitzar_resultat(equips, local, visitant, gols_local, gols_visitant)

#### actualitzar_equips():
1. Rep el dataframe df_futbol.
2. Crea el diccionari equips.
3. Crida processar_partit() per procesar cada linia (partit).
4. Retorna el diccionari equips.

In [67]:
def actualitzar_equips(df_futbol):
    """
    Rep el dataframe df_futbol.
    Retorna el diccionari equips.
    """
    equips = {}

    for _, fila in df_futbol.iterrows():
        processar_partit(equips, fila["local"], fila["visitant"], fila["gols_local"], fila["gols_visitant"])

    return equips

#### generar_classificacio():
1. Rep el diccionari equips.
2. Converteix el diccionari a dataframe df_futbol.
3. Ordena el dataframe df_futbol per punts, gols a favor i gols en contra.
4. Retorna el dataframe df_futbol.



In [68]:
def generar_classificacio(equips):
    """
    Rep el diccionari equips.
    Retorna el dataframe df_futbol ordenat.
    """
    df_futbol = pd.DataFrame.from_dict(equips, orient="index")

    df_futbol = df_futbol[["punts", "gf", "gc", "guanyats", "empatats", "perduts"]]

    df_futbol = df_futbol.sort_values(
        by=["punts", "gf", "gc"],
        ascending=[False, False, True]
    )

    return df_futbol

#### main_futbol():
1. Crida llegir_partits() per crear la llista partits.
2. Crida construir_dataframe() per crear el dataframe df_futbol.
3. Crida actualitzar_equips() per actualitzar les estadístiques de cada partit.
4. Crida generar_classificació() per generar la classificació final.
5. Retorna el dataframe df_futbol.

In [69]:
def main_futbol():
    """
    Ordena tot el procés per generar la classificació.
    Retorna el dataframe df_futbol.
    """
    partits = llegir_partits("historic_partits.txt")
    df_futbol = construir_dataframe(partits)
    equips = actualitzar_equips(df_futbol)
    df_futbol = generar_classificacio(equips)
    return df_futbol


Executa main_futbol() per obtenir el dataframe df_futbol amb la classificació.
Calcula quin és el equip més golejador.
Calcula quin és el equio més golejat.
Mostra en pantalla l'equip més golejador i l'equip més golejat.
Calcula la classificació final.
Mostra en pantalla la classificació final.
Mostra en pantalla el dataframe df_futbol.


In [70]:
df_futbol = main_futbol()

equip_mes_golejador = df_futbol["gf"].idxmax()
equip_mes_golejat = df_futbol["gc"].idxmax()

print(f"\nEquip més golejador: {equip_mes_golejador}")
print(f"Equip més golejat: {equip_mes_golejat}")
print("\nClassificació final:")

for posicio, equip in enumerate(df_futbol.index, start=1):
    print(f"{posicio}. {equip}")

df_futbol


Equip més golejador: Figueres
Equip més golejat: Vilafranca

Classificació final:
1. Girona FC
2. Llagostera
3. Sabadell
4. Cornellà
5. RCD Espanyol
6. Figueres
7. Lleida Esportiu
8. Terrassa
9. FC Barcelona
10. Vilafranca
11. Badalona
12. Nàstic de Tarragona
13. Reus Deportiu
14. Granollers
15. Olot
16. Sant Andreu
17. Manlleu
18. Cerdanyola
19. Prat
20. Europa


,punts,gf,gc,guanyats,empatats,perduts
Girona FC,96,139,94,31,3,13
Llagostera,94,159,142,29,7,20
Sabadell,85,141,121,26,7,15
Cornellà,82,147,146,25,7,22
RCD Espanyol,80,131,144,23,11,21
Figueres,79,161,154,23,10,23
Lleida Esportiu,75,129,133,23,6,23
Terrassa,74,147,164,20,14,23
FC Barcelona,73,125,115,22,7,15
Vilafranca,72,157,172,20,12,25


### Comentari sobre els resultats obtinguts
L’anàlisi automatitzada de dades esportives permet obtenir una visió clara del rendiment dels equips al llarg de la temporada i identificar patrons que serien difícils de detectar manualment. En aquest cas, però, cal destacar que la lliga analitzada presenta una particularitat important: els equips no han disputat el mateix nombre de partits. La majoria se situen entre 44 i 56 enfrontaments, amb excepcions com el Terrassa (57) i l’Europa (33). Aquesta irregularitat fa que la classificació basada únicament en punts sigui poc representativa i complica qualsevol comparativa directa entre equips.

Tot i aquest context, els resultats mostren tendències clares. El Girona FC no només lidera la classificació, sinó que ho fa amb un rendiment excepcional: ha jugat només 47 partits i presenta la mitjana de gols per partit més alta (2,95), el percentatge més elevat de victòries (65,95%) i els percentatges més baixos d’empats (6,38%) i derrotes (27,66%). A més, registra la millor diferència de gols (+45), tot i no ser l’equip més golejador (139 gols) ni el menys golejat (94 gols). Aquest conjunt d’indicadors reflecteix una gran regularitat competitiva i una eficiència notable.

Pel que fa a registres específics, el Figueres és l’equip més golejador amb 161 gols, però ocupa només la sisena posició. Això s’explica perquè també és un dels equips més golejats (154), fet que ha derivat en un nombre elevat de derrotes i empats. En l’altre extrem, l’Europa és l’equip menys golejat (93), però és últim a la classificació. La seva situació respon principalment a dos factors: és l’equip amb menys partits disputats (33) i ha marcat exactament els mateixos gols que ha rebut, cosa que limita la seva capacitat de sumar punts.

En conjunt, aquesta automatització facilita la gestió de dades històriques i permet obtenir insights rellevants sobre el rendiment dels equips, detectar desequilibris tàctics i identificar àrees de millora. També posa de manifest la importància de normalitzar les dades quan la competició no és homogènia, especialment si es volen extreure conclusions comparables o permetre anticipar patrons de rendiment esportiu per a temporades futures.